### ISSUE TITLE
[Question] Time limits

### ISSUE THREAD

**Issue by CorentinJ** [montefiore-ai/alan-cluster#9]
Is there any imposed time limit on all jobs, or one that is agreed by everyone? I've not seen jobs go beyond 24h on the cluster. I have three different models that must be (consecutively) trained for at least one week each. Is it alright if I set the time limit to seven days?

**Comment by JoeriHermans** [comment]
Hi Corentin,

The following time-limits are imposed:
- 10 minutes by default (if no `--time` flag has been specified).
- Max 7 days if `--time` is specified.

So you will be able to train for a week, if desired.

**Comment by JoeriHermans** [comment]
@CorentinJ could you check your GPU utilization, it seems to be very low (0-5%).

Joeri

**Comment by CorentinJ** [comment]
Thank you for reminding me. I'm assuming you're referring to the GPU memory usage. In the task I've implemented, the number of operations for computing the loss grows in a cubic fashion with respect to the batch size, so I cannot set a batch size too high. I planned to look up how to parallelize this task on the same GPU (provided that I'm not bound by the data loader I/O) and I forgot about it. I'll see to it, but either way I'm almost done training that network.

By the way, could you tell me how to monitor the GPU usage?

**Comment by JoeriHermans** [comment]
> I'm assuming you're referring to the GPU memory usage.
Not really, I'm talking about the "compute-utilization".

You can do the following to monitor the utilization of your job. Check what node your job is running on, in this case `alan-compute-05`, and then do `nvidia-smi`.

**Comment by CorentinJ** [comment]
Thank you. Well, I've noticed that my average step execution time went up by ~40% from when it was running on eddard, but it's hard to diagnose what's going on like this.

**Comment by JoeriHermans** [comment]
> went up by ~40%
That's good right? You are probably using TensorFlow I guess? So I cannot really help with that.

**Comment by CorentinJ** [comment]
No I'm on pytorch actually, and it's the step execution *time* so it's rather bad :/

My guess is that it's either
- I/O bound, if the disk is much slower and that the data loading in the other threads cannot match the step execution time.
- CPU bound, and either the computation of the loss (done on the CPU, it's faster than on the GPU) or the threads loading the data are the bottleneck.

I would have to log the execution time of these operations as well, I can do that on my next run.

**Comment by JoeriHermans** [comment]
Are you using asynchronous data-loaders with multiple works? Did you set `pin_memory=True`? Because your processes look super-inactive tbh. I'm not sure were the bottleneck is.

**Comment by CorentinJ** [comment]
I'm using `torch.utils.data.Dataset` and `torch.utils.data.DataLoader` with 4 workers. I haven't set pin_memory to True because I have a custom collate function and I don't know if they mix well together. 

```
Average execution time over 10 steps:
  Blocking, waiting for batch (threaded) (10/10):  mean:   61ms   std:   14ms
  Data to cuda (10/10):                            mean:    2ms   std:    0ms
  Forward pass (10/10):                            mean:   30ms   std:    1ms
  Loss (10/10):                                    mean:  325ms   std:    4ms
  Backward pass (10/10):                           mean: 1353ms   std:   29ms
  Parameter update (10/10):                        mean:   86ms   std:    2ms
  Extras (visualizations, saving) (10/10):         mean:   12ms   std:    0ms
```

But either way, I've set up profiling on my local machine (running a GTX1080) and you can see that the bottleneck really is at the backward pass. As you can see, the loss is 10x slower than the forward pass, because it is cubic (with respect to the batch size) in complexity. The backward pass takes about 4 times the duration of the forward pass + loss, which is reasonable (IIRC a good rule of thumb is 3 times the forward pass + loss). I'm no expert, but I think that the backward pass is strictly done on the CPU and thus I cannot benefit from much of a speedup through parallelism.

Note that computing the loss on the GPU is much slower:
```
Average execution time over 10 steps:
  Blocking, waiting for batch (threaded) (10/10):  mean:   61ms   std:   11ms
  Data to cuda (10/10):                            mean:    2ms   std:    0ms
  Forward pass (10/10):                            mean:    5ms   std:    0ms
  Loss (10/10):                                    mean: 1870ms   std:   39ms
  Backward pass (10/10):                           mean: 5060ms   std:   97ms
  Parameter update (10/10):                        mean:    3ms   std:    0ms
  Extras (visualizations, saving) (10/10):         mean:   12ms   std:    1ms
```

**Comment by JoeriHermans** [comment]
Are you sure you moved your model to the GPU - before - allocating the optimizer?

**Comment by CorentinJ** [comment]
Yep:

```
    # Create the model and the optimizer
    model = Model(hparams).cuda()
    optimizer = torch.optim.Adam(model.parameters(), lr=hparams.learning_rate_init)
```

If you're very motivated, I can share the source code :D

**Comment by CorentinJ** [comment]
I want to add that I run this on my local computer and let it take 100% of my combined CPU usage. When I ran it on the cluster, I think I specified 1 CPU per task (or left it as default), so I'm fairly confident this is the source of the slowdown. Sorry for bringing this up just now. As for the GPU usage, it will probably remain somewhat low even with more CPUs.

**Comment by CorentinJ** [comment]
Here is the execution on the cluster with 4 CPUs:

```
Average execution time over 10 steps:                                         
  Blocking, waiting for batch (threaded) (10/10):  mean:  189ms   std:  207ms 
  Data to cuda (10/10):                            mean:    2ms   std:    1ms 
  Forward pass (10/10):                            mean:   11ms   std:    3ms 
  Loss (10/10):                                    mean:  402ms   std:   16ms 
  Backward pass (10/10):                           mean: 2321ms   std:  205ms 
  Parameter update (10/10):                        mean:   55ms   std:    1ms 
  Extras (visualizations, saving) (10/10):         mean:   44ms   std:    5ms 
```

So it's slower than on my local computer, but I'm not sure why. Could be the CPU, I have an i5-8600k . Here is my script:
```
#!/bin/bash
#SBATCH --cpus-per-task=4
#SBATCH --mem=4G
#SBATCH --gres=gpu:1
#SBATCH --time=7-00:00:00
#SBATCH --output="output.txt"
python encoder_train.py pretrained /home/cjemine/datasets/SV2TTS/encoder --visdom_server="alan-master"
```

**Comment by JoeriHermans** [comment]
Looking at your loss, I would recommend to vectorize it. You do note take any advantage here of GPU parallelization, and compute the tensor elements 1-by-1. This is the source of your bottleneck, and why the loss computation (and the accompanying backward pass) takes so long.

**Comment by CorentinJ** [comment]
Well shit, thanks Joeri. I knew that triple for loop was problematic but I never managed to wrap my head around how to parallelize it... You gave me the motivation to think it through again, and now it's much better. I'm still CPU bound, but I got a 4.7x speedup on my step execution time (my computer):

```
Average execution time over 10 steps:
  Blocking, waiting for batch (threaded) (10/10):  mean:   64ms   std:   10ms
  Data to cuda (10/10):                            mean:    2ms   std:    0ms
  Forward pass (10/10):                            mean:   30ms   std:    0ms
  Loss (10/10):                                    mean:   48ms   std:    1ms
  Backward pass (10/10):                           mean:  163ms   std:    4ms
  Parameter update (10/10):                        mean:   87ms   std:    2ms
  Extras (visualizations, saving) (10/10):         mean:   12ms   std:    0ms
```

**Comment by CorentinJ** [comment]
I've looked more into it, the bottleneck was definitely the file I/O. Since it is asynchronous with the training, the loss was the bigger bottleneck. Now that it's fixed, it's clear that there is a problem on the end of the data loader.

Compare the execution on the cluster:
```
Average execution time over 10 steps:                                         
  Blocking, waiting for batch (threaded) (10/10):  mean: 7824ms   std: 14883ms
```
with the execution on my computer:
```
Average execution time over 10 steps:
  Blocking, waiting for batch (threaded) (10/10):  mean:   87ms   std:   63ms
```

My dataset is split in a lot of small files (numpy array and text files), so maybe there's a stronger limit on the number of files opened at once on the cluster? When I look deeper into what operations are causing the slowdown, I see that it's always reading from a text file or using `numpy.load`. All these I/O operations take less than one millisecond on my end, but on the cluster I see that e.g. reading 100 short lines from a file sometimes takes 150ms.

**Comment by glouppe** [comment]
If you read/write files from your home directory, you are actually going through the network file system. You might be hitting the issue because of that.

To speed things up, you can instead start moving the files you will use in `/scratch/` (which is local to the node), and then doing your I/O from there. This should help.

**Comment by CorentinJ** [comment]
Thanks.
- Is it ok to put a large dataset in there? (~250gb)
- I assume the /scratch/ is not shared between nodes? If that's the case, I would have to know in advance which node my job will be run on. How can I do that?

**Comment by glouppe** [comment]
Yes it is fine. The scratch will be flushed every now and then. (@JoeriHermans not sure what period you set?)

You can include the copying operations directly inside your slurm scripts. In this way, you do not have to know in advance on which node your script will run.

**Comment by CorentinJ** [comment]
I do not have permission to write on /scratch/, they are set to `rwxr-xr-x`.

**Comment by glouppe** [comment]
Indeed, I was a bit too fast on that. This is still on our TODO list.

Gilles

**Comment by CorentinJ** [comment]
What's the role of `/data/datasets` by the way? Is it also accessible only through the network for the compute nodes?

I'm going to run into the same problem for my two other models that I have left to train, all my datasets are big. If the bandwidth is good then it's probably ok if I copy the entirety of them for training every time. I can also use `-w alan-compute-x` to remain on the same node when debugging,

**Comment by JoeriHermans** [comment]
Yes, still have to setup the proper Ansible scripts for configuring the scratch. Also have to add some additional documentation since the scratch is quite limited, i.e., users should delete their data after the job is complete.

Alternatively, we could do it like this: when a job is allocated, Slurm creates the scratch directory for them and deletes the directory when it is completed. I personally think this is the way to do it, but I need to check the Slurm docs if we can hook some custom code in there.

The purpose of the `/data/datasets/` directory is to store commonly used datasets (e.g., ImageNet, CIFAR-10 etc) to minimize the copies of those datasets on the cluster.

**Comment by CorentinJ** [comment]
But these datasets are still remote to the compute nodes I presume?

**Comment by JoeriHermans** [comment]
Indeed. I'm making the necessary configuration changes now to enable the following behaviour:

> Alternatively, we could do it like this: when a job is allocated, Slurm creates the scratch directory for them and deletes the directory when it is completed. I personally think this is the way to do it, but I need to check the Slurm docs if we can hook some custom code in there.

Should be up in an hour. Will ping you when it is ready.
